src/main/java/org/opensearch/knn/index/memory/NativeMemoryCacheManager.java

```diff
@@ -0,0 +1,281 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.memory;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheStats;
+import com.google.common.cache.RemovalCause;
+import com.google.common.cache.RemovalNotification;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.knn.common.exception.OutOfNativeMemoryException;
+import org.opensearch.knn.index.KNNSettings;
+import org.opensearch.knn.plugin.stats.StatNames;
+
+import java.io.Closeable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Manages native memory allocations made by JNI.
+ */
+public final class NativeMemoryCacheManager implements Closeable {
+
+    public static String GRAPH_COUNT = "graph_count";
+
+    private static Logger logger = LogManager.getLogger(NativeMemoryCacheManager.class);
+    private static NativeMemoryCacheManager INSTANCE;
+
+    private Cache<String, NativeMemoryAllocation> cache;
+    private ExecutorService executor;
+    private AtomicBoolean cacheCapacityReached;
+    private long maxWeight;
+
+    NativeMemoryCacheManager() {
+        this.executor = Executors.newSingleThreadExecutor();
+        this.cacheCapacityReached = new AtomicBoolean(false);
+        this.maxWeight = Long.MAX_VALUE;
+        initialize();
+    }
+
+    /**
+     * Make sure we just have one instance of cache.
+     *
+     * @return NativeMemoryCacheManager instance
+     */
+    public static synchronized NativeMemoryCacheManager getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new NativeMemoryCacheManager();
+        }
+        return INSTANCE;
+    }
+
+    private void initialize() {
+        CacheBuilder<String, NativeMemoryAllocation> cacheBuilder = CacheBuilder.newBuilder()
+                .recordStats()
+                .concurrencyLevel(1)
+                .removalListener(this::onRemoval);
+
+        if(KNNSettings.state().getSettingValue(KNNSettings.KNN_MEMORY_CIRCUIT_BREAKER_ENABLED)) {
+            maxWeight = KNNSettings.getCircuitBreakerLimit().getKb();
+            cacheBuilder.maximumWeight(maxWeight).weigher((k, v) -> (int)v.getSizeInKb());
```

Comment:
if KNN_MEMORY_CIRCUIT_BREAKER_ENABLED is disabled, do you have other eviction trigger mechanism?

Reply:
User can set time based eviction policy. I would like to eventually add a clear cache API.